Add PSQL ODBC tests for XML datatype

rossd-bitquill opened the pull request:
### Description

This PR adds tests for the PSQL XML datatype including testing:
Table Creation
Insertion Success
Update Success
View Creation

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

### Issues Resolved
BABELFISH-563

### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Add PSQL ODBC tests for XML datatype:
Table Creation
Insertion Success
Update Success
Data Length

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Add test for View creation with XML

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Fix comments, remove TODO

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Add larger XML test

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Remove test for insertion failure

Test was giving unusual results, investigating further before commiting
that test.

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+    NORMAL_XML,
+    LARGE_XML,
+  };
+  const int number_of_entries = inserted_values.size();
```
```suggestion
  const int NUMBER_OF_ENTRIES = inserted_values.size();
```

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+    NORMAL_XML,
+    LARGE_XML
+  };
+  const int number_of_entries = inserted_values.size();
```
```suggestion
  const int NUMBER_OF_ENTRIES = inserted_values.size();
```

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+};
+
+const string SIMPLE_XML = "<root>\
+<book>-148708948</book>\
+<leaf>simple</leaf>\
+<remain>shake</remain>\
+<football>nest</football>\
+<blanket>1275892738</blanket>\
+<try>-249153758.16992378</try>\
+</root>";
```
```suggestion
  <book>-148708948</book>\
  <leaf>simple</leaf>\
  <remain>shake</remain>\
  <football>nest</football>\
  <blanket>1275892738</blanket>\
  <try>-249153758.16992378</try>\
  </root>";
```

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+</root>";
+
+const string NORMAL_XML = "<root>\
+<that>hot</that>\
+<anyone>1386369263.1890697</anyone>\
+<finest>believed</finest>\
+<song>\
+<stems>\
+<headed>1563641070.153514</headed>\
+<bring>-1175941901.956747</bring>\
+<stand>-494359546.04893684</stand>\
+<pole>774686962</pole>\
+<outer>production</outer>\
+<why>applied</why>\
+</stems>\
+<breeze>1269756220</breeze>\
+<cookies>544666062</cookies>\
+<yesterday>grade</yesterday>\
+<power>-663154896.7426605</power>\
+<topic>stomach</topic>\
+</song>\
+<fact>-1834471409</fact>\
+<repeat>-1980145795</repeat>\
+</root>";
```
```suggestion
  <that>hot</that>\
  <anyone>1386369263.1890697</anyone>\
  <finest>believed</finest>\
  <song>\
  <stems>\
  <headed>1563641070.153514</headed>\
  <bring>-1175941901.956747</bring>\
  <stand>-494359546.04893684</stand>\
  <pole>774686962</pole>\
  <outer>production</outer>\
  <why>applied</why>\
  </stems>\
  <breeze>1269756220</breeze>\
  <cookies>544666062</cookies>\
  <yesterday>grade</yesterday>\
  <power>-663154896.7426605</power>\
  <topic>stomach</topic>\
  </song>\
  <fact>-1834471409</fact>\
  <repeat>-1980145795</repeat>\
  </root>";
```

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+</root>";
+
+const string LARGE_XML = "<Root>\
+ <sys.tables name=\"tnullfixeddecimal0100\" object_id=\"4195065\" schema_id=\"1\" parent_object_id=\"0\" type=\"U \" type_desc=\"USER_TABLE\" create_date=\"2022-02-03T19:23:37.030\" modify_date=\"2022-02-03T19:23:37.030\" is_ms_shipped=\"0\" is_published=\"0\" is_schema_published=\"0\" lob_data_space_id=\"0\" max_column_id_used=\"5\" lock_on_bulk_load=\"0\" uses_ansi_nulls=\"1\" is_replicated=\"0 ...
```
Nit: This is indented by 1 space. Can you fix indentation by indenting to 2 spaces?

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+  const vector<string> BBF_NAME_EXPECTED = {"int", "xml"};
+
+  testCommonColumnAttributes(ServerType::MSSQL, BBF_TABLE_NAME, 
+      TABLE_COLUMNS.size(), COL1_NAME, 
+      BBF_LENGTH_EXPECTED, BBF_PRECISION_EXPECTED, 
+      BBF_SCALE_EXPECTED, BBF_NAME_EXPECTED);
```
```suggestion
    TABLE_COLUMNS.size(), COL1_NAME, 
    BBF_LENGTH_EXPECTED, BBF_PRECISION_EXPECTED, 
    BBF_SCALE_EXPECTED, BBF_NAME_EXPECTED);
```

avneet-toor-bq commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+  dropObject(ServerType::PSQL, "VIEW", PG_VIEW_NAME);
+  dropObject(ServerType::MSSQL, "TABLE", BBF_TABLE_NAME);
+  dropObject(ServerType::PSQL, "TABLE", PG_TABLE_NAME);
+}
```
Nit: Please add new line at end of file.

rossd-bitquill replied:
Done.

Commit:
Address review comments

Formatting only.

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Merge https://github.com/babelfish-for-postgresql/babelfish_extensions into JIRA-BABELFISH-563

ColinKYuen commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+  dropObject(ServerType::MSSQL, "TABLE", BBF_TABLE_NAME);
+}
+
+TEST_F(PSQL_DataTypes_Xml, Data_Length) {
```
What's the difference between this test and `Insertion_Success`

rossd-bitquill replied:
Excellent question.  I had a different version of this at one point, but it wasn't working so I thought I removed it, but I guess I didn't.

Will remove it.

ColinKYuen commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,230 @@
+  testInsertionSuccess(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, inserted_values, inserted_values);
+
+  testUpdateSuccess(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, COL2_NAME, data_update_values, data_update_values);
+  verifyValuesInObject(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, data_update_values, data_update_values);
+
+  data_update_values = {
+    SIMPLE_XML
+  };
```
```suggestion
  verifyValuesInObject(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, data_update_values, data_update_values);
```

rossd-bitquill replied:
The whole purpose of putting something different was to test that the update actually happened with different data, but when I factored the XML text out into file-level constants I forgot that, so changed instead to LARGE_XML to have a different value update test.

Commit:
Remove duplicated test case

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Merge https://github.com/babelfish-for-postgresql/babelfish_extensions into JIRA-BABELFISH-563

Commit:
Update tests with new data

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

ColinKYuen commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,207 @@
+
+  testInsertionSuccess(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, inserted_values, inserted_values);
+
+  testUpdateSuccess(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, COL2_NAME, data_update_values, data_update_values);
+  verifyValuesInObject(ServerType::MSSQL, BBF_TABLE_NAME, COL1_NAME, data_update_values, data_update_values);
```
Sorry, didn't catch this earlier.

`testUpdateSuccess` already does something similar to `verifyValuesInObject`

`verifyValuesInObject` should be used on the other connection
ie After updating with PG connection, `verifyValuesInObject` on BBF

rossd-bitquill replied:
Ah, that makes sense, so I'll switch the two calls to varifyValuesInObject to refer to the other connection type.

Commit:
Fix Table Update tests

Table Update tests were calling VerifyObjects on the wrong connection,
basically just re-doing work already verified.  Fix that by switch the
connection verified.

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Fix failing tests

Queried the wrong table for verification.

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

ColinKYuen commented on test/odbc/psqlodbc/xml.cpp:
```diff
@@ -0,0 +1,207 @@
+#include "psqlodbc_tests_common.h"
+#include "../src/query_generator.h"
```
```suggestion
```

Commit:
Remove unneeded import

Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

Commit:
Merge https://github.com/babelfish-for-postgresql/babelfish_extensions into JIRA-BABELFISH-563
